### PR TITLE
[major] Change the minimum width of the result of "Shift Right" for UInt to 0-bit

### PR DESCRIPTION
**Pull request by SpriteOvO**
Change the minimum width of the result of "Shift Right" to 0-bit to align the behavior between Chisel and CIRCT.

Context: https://github.com/chipsalliance/chisel/pull/3752#issuecomment-1905086931, https://github.com/chipsalliance/chisel/pull/3752#issuecomment-1906225248 and chipsalliance/chisel#3735.

### CODE REVIEW

**Comment by dtzSiFive** [comment]
The text that follows says:

> If n is greater than or equal to the bit-width of e, the resulting value will be zero for unsigned types and the sign bit for signed types.
n must be non-negative.

Which suggests maybe the table should still have `1` for the `SInt` case?

Consider:

```
circuit SIntShr:
  module SIntShr:
    input in: SInt<5>
    output out: SInt
    out <= shr(in, 8)
```

Which feeding through `firtool` produces:

```
// Generated by CIRCT firtool-1.63.0g20240122_ce27439
module SIntShr(
  input  [4:0] in,
  output       out
);

  assign out = in[4];
endmodule
```

**Comment by jackkoenig** [comment]
Good catch @dtzSiFive, I agree. The analogy with `tail` is good, but `tail` on an `SInt` returns `UInt` whereas `shr` on an `SInt` returns `SInt` and as such should preserve the sign bit.

File: revision-history.yaml [hunk 1]
```diff
@@ -13,6 +13,7 @@ revisionHistory:
       - Reorganized statements section.
       - Rewrite of the Types section.
       - Fix mistake in layer-associated probe grammar.
+      - Change the minimum width of the result of "Shift Right" to 0-bit.
```
Comment (jackkoenig) [inline revision-history.yaml:16]:
```suggestion
      - Change the minimum width of the result of "Shift Right" on UInt to 0-bit.
```

File: spec.md [hunk 2]
```diff
@@ -2999,9 +2999,9 @@ n must be non-negative.
   --------------------------------------------------------------------------
   Name   Arguments   Parameters   Arg Types   Result Type   Result Width
   ------ ----------- ------------ ----------- ------------- ----------------
-  shr    \(e\)       \(n\)        (UInt)      UInt          max(w~e~-n, 1)
+  shr    \(e\)       \(n\)        (UInt)      UInt          max(w~e~-n, 0)
 
-                                  (SInt)      SInt          max(w~e~-n, 1)
+                                  (SInt)      SInt          max(w~e~-n, 0)
```
Comment (jackkoenig) [inline spec.md:3004]:
```suggestion
                                  (SInt)      SInt          max(w~e~-n, 1)
```

**Review by jackkoenig — CHANGES_REQUESTED** [review]
After offline discussion with @darthscsi, @seldridge, and @dtzSiFive, we think the best course of action is to change just UInt to have a minimum width of 0 and keep SInt with a minimum width of 1 to maintain mathematical soundness with what it means to do an arithmetic shift to a signed number.

**Comment by SpriteOvO** [comment]
> we think the best course of action is to change just UInt to have a minimum width of 0 and keep SInt with a minimum width of 1 to maintain mathematical soundness with what it means to do an arithmetic shift to a signed number.

Makes sense to me, changes applied.

**Review by jackkoenig — APPROVED** [review]

**Review by seldridge — COMMENTED** [review]
Thanks for working with us to figure out the right thing to do here. 👍

**Pull request merged**